### PR TITLE
[#16] user based cross device tracking

### PR DESCRIPTION
**Pull request by gingerlime**
* adding an optional user_id attribute to experiments
* when this attribute is present, variant selection will
  be based on the user_id as part of the seed for a hash
  function (along with experiment name and the type of caller)
* tracking uuid for Gimel and Keen will also change to be
  based on the user_id as seed, so the same user will only
  get tracked once from different devices. Unless the goal
  is not-unique.
* updated tests and added adapter tests for Gimel and Keen.io

### CODE REVIEW

File: src/alephbet.coffee [hunk 1]
```diff
@@ -59,25 +61,30 @@ class AlephBet
...
       return active unless typeof active is 'undefined'
-      active = Math.random() <= @options.sample
+      active = @_random('sample') <= @options.sample
       @storage().set("#{@options.name}:in_sample", active)
       active
 
+    _random: (salt) ->
+      return utils.random() unless @user_id
+      seed = "#{@name}.#{salt}.#{@user_id}"
+      utils.random(seed)
```
Comment (joker-777) [inline src/alephbet.coffee]: Why do you need this seed for the `user_id`? Can't you just always use ` utils.random("#{@name}.#{salt}.#{@user_id}")`?

File: CHANGES [hunk 2]
```diff
@@ -1,3 +1,12 @@
+* 0.15.0-rc.1 BREAKING changes (only if you wrote your own tracking adapter)
+    - User-based / Cross-device tracking
+    - If you use your own tracking adapter, you'll have to change it.
+      * `experiment_start` now accepts the `experiment` as the first parameter
```
Comment (gingerlime) [inline CHANGES]: Yes, great suggestion!

File: README.md [hunk 3]
```diff
@@ -197,11 +232,11 @@ Here's an example for integrating an adapter for [keen.io](https://keen.io)
...
     });
     var tracking_adapter = {
-        experiment_start: function(experiment_name, variant) {
-            keen_client.addEvent(experiment_name, {variant: variant, event: 'participate'});
+        experiment_start: function(experiment, variant) {
+            keen_client.addEvent(experiment.name, {variant: variant, event: 'participate'});
         },
-        goal_complete: function(experiment_name, variant, event_name) {
-            keen_client.addEvent(experiment_name, {variant: variant, event: event_name});
+        goal_complete: function(experiment, variant, event_name, _props) {
```
Comment (gingerlime) [inline README.md]: by convention variables that are passed over but not used are marked with `_` prefix. I think it's useful to know that it's being passed over from the experiment, even if it's not used.

File: src/adapters.coffee [hunk 4]
```diff
@@ -52,28 +52,37 @@ class Adapters
 
     _flush: ->
       for item in @_queue
-        callback = @_remove_uuid(item.properties.uuid)
-        @_ajax_get(@url, item.properties, callback)
+        callback = @_remove_quuid(item.properties._quuid)
+        @_ajax_get(@url, utils.omit(item.properties, '_quuid'), callback)
```
Comment (gingerlime) [inline src/adapters.coffee]: there's no need to send it over the wire.

File: src/adapters.coffee [hunk 5]
```diff
@@ -157,11 +175,11 @@ class Adapters
       throw 'ga not defined. Please make sure your Universal analytics is set up correctly' if typeof ga isnt 'function'
       ga('send', 'event', category, action, label, {'nonInteraction': 1})
 
-    @experiment_start: (experiment_name, variant) =>
-      @_track(@namespace, "#{experiment_name} | #{variant}", 'Visitors')
+    @experiment_start: (experiment, variant) =>
+      @_track(@namespace, "#{experiment.name} | #{variant}", 'Visitors')
```
Comment (gingerlime) [inline src/adapters.coffee]: not in this tracking adapter... the tracking adapters don't share any code and can implement internally differently.

File: src/adapters.coffee [hunk 6]
```diff
@@ -157,11 +175,11 @@ class Adapters
...
 
-    @goal_complete: (experiment_name, variant, goal) =>
-      @_track(@namespace, "#{experiment_name} | #{variant}", goal)
+    @goal_complete: (experiment, variant, goal, _props) =>
```
Comment (gingerlime) [inline src/adapters.coffee]: Thanks for spotting. Will change this.

File: src/alephbet.coffee (hunk 1 above)
Comment (gingerlime) [inline src/alephbet.coffee]: Are you asking why I'm assigning it to a variable? not strictly necessary. I can pass it directly as you suggested. But might be a bit more readable to understand what this string is used for?

File: src/adapters.coffee (hunk 5 above)
Comment (joker-777) [inline src/adapters.coffee]: In general I think you should try to keep the interfaces of the adapters similar.

File: src/alephbet.coffee (hunk 1 above)
Comment (joker-777) [inline src/alephbet.coffee]: No that wasn't my question. My question is why you can't use `utils.random("#{@name}.#{salt}.#{@user_id}")` even when there is no `@user_id`? Simply writing

```coffee
_random: ->
  utils.random("#{@name}.#{@user_id}")
```

File: src/adapters.coffee (hunk 5 above)
Comment (gingerlime) [inline src/adapters.coffee]: The interface is the same. This is internal implementation.

File: src/alephbet.coffee (hunk 1 above)
Comment (gingerlime) [inline src/alephbet.coffee]: That won't be random though...

Comment (joker-777) [inline src/alephbet.coffee]: Can you quickly explain how does it work with the salt?

Comment (gingerlime) [inline src/alephbet.coffee]: Take a look at the implementation. It's super small and simple. It uses a hash function with the seed and generates a random number based on the seed. If the seed is different for different users, the result hash will produce a different hash, which would get "translated" to a different number between 0 and 1. So for a given experiment, salt and user_id, we'll get a pseudo-random number. If you send the same user_id (or `null`), then you'll always get the same result, which won't be (pseudo)random...

Comment (gingerlime) [inline src/alephbet.coffee]: When we don't have a user_id, using a "real" random number is preferable.

Comment (gingerlime) [inline src/alephbet.coffee]: #16 has a link to [planout](https://facebook.github.io/planout/docs/how-planout-works.html) which might explain this in more detail.

Comment (joker-777) [inline src/alephbet.coffee]: ok, thanks

**Comment by joker-777** [comment]
@gingerlime Somehow I missed it. Why do I need to specifically add the `user_id` check to the goals and also to the triggers. Can't you do it internally? Is there a use case where you add the `user_id` but want to trigger an experiment even if it doesn't exist?

**Comment by gingerlime** [comment]
@joker-777 I wrote about it on this wiki page https://github.com/Alephbet/alephbet/wiki/User-based-and-Cross-device-tracking

>  Is there a use case where you add the user_id but want to trigger an experiment even if it doesn't exist?

I don't think you can mix not-logged-in and logged-in on the same experiment, without getting unreliable results unfortunately. I tried to cover a couple of cases where this might cause inconsistent results on the wiki page.

> Why do I need to specifically add the user_id check to the goals and also to the triggers. Can't you do it internally?

It might be possible to do this internally, but this definitely complicates things. You'd need to store/cache the user_id. You'd need to know when the user_id is missing, vs when it wasn't meant to be used at all... It's probably a good idea, but need to think carefully and design it to work seamlessly without creating unexpected behaviour. For now, I think it's ok to expect the caller to make sure the user_id is available, and it's not too difficult to cache it externally if you want to "remember" who the user was initially.

**Comment by joker-777** [comment]
The only way to achieve this is to add another attribute, in addition to `user_id`. Something like `cross_device: true`. I don't think it is to complicated to add some small checks internally. In fact I think you only need to add `return if @cross_device && !@user_id` to the method which calls the `trigger` callback and to `add_goal` and `add_goals` methods. I didn't dig into the code though.

**Comment by gingerlime** [comment]
two attributes for one purpose makes the Alephbet API ugly. If the intention is to make it transparent, then this should be handled internally with only the `user_id` attribute. It does make things a bit more complicated however.

**Pull request merged**